### PR TITLE
REWARDS: Protocol wallet (staking, bootstrap)

### PR DESCRIPTION
**Pull request by vistra**

### CODE REVIEW

File: contracts/ProtocolWallet.sol [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    /// @dev transfer the entire pool's balance to a new wallet.
+    function emergencyWithdraw() external onlyMigrationOwner {
+        emit EmergencyWithdrawal(msg.sender);
+        require(token.transfer(msg.sender, getBalance()), "ProtocolWallet::emergencyWithdraw - transfer failed");
+    }
+
+    /// @dev sets the address of the new contract
+    function setClient(address newClient) external onlyFunctionalOwner {
+        client = newClient;
```
[the remark this answers is not in the document]
Comment (vistra) [reply contracts/ProtocolWallet.sol:76]: I don't see why not. It's revertable.

**Comment by vistra** [comment]
1. Not sure about style, but using `msg.sender` is cheaper in gas.
2. Good point. I don't think there is a practical issue here though - the rate can be slightly adjusted upwards to handle that. Besides, we don't plan this to be accurate to the Orbiton. From what I understand the plan is to allow a significant margin of error, say 20%. 
3.  By overflow I assume you mean exceeding the allowed rate, not int overflow, right? IDK, letting the calling contract handle a lower-than-expected balance sounds complicated. Not sure what we would benefit from it. 
As for atomically decreasing the rate, we can still do that non-atomically in our scenarios (bootstrap and staking rewards) if we first adjust the rewards rate and then the wallet rate. Am I missing something?

**Comment by ronnno** [comment]
1. I meant that more as a question. 
2. X% margin as a policy is satisfactory. But this pitfall stands out and it begged the question. should be explicit about it. Are we ok product wise  with allowing more than we Actually pay?

3. Not relevant if we are prepared to allow a X% margin.

File: contracts/ProtocolWallet.sol [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    using SafeMath for uint256;
+
+    IERC20 public token;
+    address public client;
+
+    uint lastApprovedAt;
+    uint annualRate;
+
+    modifier onlyClient() {
+        require(msg.sender == client, "caller is not the committee contract");
```
Comment (OdedWx) [inline contracts/ProtocolWallet.sol:18]: caller is not the wallet client.
BTW, why committee?

**Comment by vistra** [comment]
Right, c&p mistake :)

File: contracts/ProtocolWallet.sol [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+        _;
+    }
+
+    constructor(IERC20 _token, address _client) public {
+        token = _token;
+        client = _client;
+        lastApprovedAt = now;
```
Comment (OdedWx) [inline contracts/ProtocolWallet.sol:26]: Makes sense to init in the constructor, an alternative is to init on setMaxAnnualRate when set from 0 to != 0.
Need to verify that the client can operate correctly with 0 value (ERC20 does) to allow early deployment.

File: contracts/ProtocolWallet.sol [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        uint maxAmount = duration.mul(annualRate).div(365 * 24 * 60 * 60);
+        require(amount <= maxAmount, "ProtocolWallet:approve - requested amount is larger than allowed by rate");
+
+        lastApprovedAt = now;
+        require(token.transfer(msg.sender, amount), "ProtocolWallet::withdraw - transfer failed");
+    }
+
+    /* Governance */
+    /// @dev Sets a new transfer rate for the Orbs pool.
+    function setMaxAnnualRate(uint256 annual_rate) external onlyMigrationOwner {
```
Comment (OdedWx) [inline contracts/ProtocolWallet.sol:63]: it's common to use _annualRate

File: contracts/ProtocolWallet.sol [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    /// Flow:
+    /// PoolWallet.approveTransfer(amount);
+    /// ERC20.transferFrom(PoolWallet, client, amount)
+    function withdraw(uint256 amount) external onlyClient {
+        uint duration = now - lastApprovedAt;
+        uint maxAmount = duration.mul(annualRate).div(365 * 24 * 60 * 60);
+        require(amount <= maxAmount, "ProtocolWallet:approve - requested amount is larger than allowed by rate");
+
+        lastApprovedAt = now;
+        require(token.transfer(msg.sender, amount), "ProtocolWallet::withdraw - transfer failed");
```
Comment (OdedWx) [inline contracts/ProtocolWallet.sol:58]: //TODO May skip the transfer on amount == 0.

File: contracts/ProtocolWallet.sol [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    /// @dev transfer the entire pool's balance to a new wallet.
+    function emergencyWithdraw() external onlyMigrationOwner {
+        emit EmergencyWithdrawal(msg.sender);
+        require(token.transfer(msg.sender, getBalance()), "ProtocolWallet::emergencyWithdraw - transfer failed");
+    }
+
+    /// @dev sets the address of the new contract
+    function setClient(address newClient) external onlyFunctionalOwner {
```
Comment (OdedWx) [inline contracts/ProtocolWallet.sol:75]: _client

**Comment by OdedWx** [comment]
2# is a good point, we don't plan +20% initially, but may do so in the future to allow flexibility with the client contract governance. one can still add a margin.
A simple way to round up is: n+(m-1)/m. 
uint maxAmount = duration.mul(annualRate + 365 days - 1).div(365 days);

3# TBD, it may be better to identify an error than leave the system in an unexpected state.

**Pull request merged**